File: cli/internal/nodes/packagetask.go
// Package nodes defines the nodes that are present in the execution graph used by turbo.
package nodes

import (
	"fmt"
	"path/filepath"
	"strings"

	"github.com/vercel/turbo/cli/internal/fs"
	"github.com/vercel/turbo/cli/internal/fs/hash"
	"github.com/vercel/turbo/cli/internal/util"
)

// PackageTask represents running a particular task in a particular package
type PackageTask struct {
	TaskID          string
	Task            string
	PackageName     string
	Pkg             *fs.PackageJSON
	EnvMode         util.EnvMode
	TaskDefinition  *fs.TaskDefinition
	Dir             string
	Command         string
	Outputs         []string
	ExcludedOutputs []string
	Hash            string
}

const logDir = ".turbo"

// RepoRelativeSystemLogFile returns the path from the repo root
// to the log file in system format
func (pt *PackageTask) RepoRelativeSystemLogFile() string {
	return filepath.Join(pt.Dir, logDir, logFilename(pt.Task))
}

func (pt *PackageTask) packageRelativeSharableLogFile() string {
	return strings.Join([]string{logDir, logFilename(pt.Task)}, "/")
}

func logFilename(taskName string) string {
	escapedTaskName := strings.ReplaceAll(taskName, ":", "$colon$")
	return fmt.Sprintf("turbo-%v.log", escapedTaskName)
}

// OutputPrefix returns the prefix to be used for logging and ui for this task
func (pt *PackageTask) OutputPrefix(isSinglePackage bool) string {
	if isSinglePackage {
		return pt.Task
	}
	return fmt.Sprintf("%v:%v", pt.PackageName, pt.Task)
}

// HashableOutputs returns the package-relative globs for files to be considered outputs
// of this task
func (pt *PackageTask) HashableOutputs() hash.TaskOutputs {
	inclusionOutputs := []string{pt.packageRelativeSharableLogFile()}
	inclusionOutputs = append(inclusionOutputs, pt.TaskDefinition.Outputs.Inclusions...)

	hashable := hash.TaskOutputs{
		Inclusions: inclusionOutputs,
		Exclusions: pt.TaskDefinition.Outputs.Exclusions,
	}
	hashable.Sort()
	return hashable
}

File: crates/turborepo-lib/src/package_task.rs
use turbopath::AbsoluteSystemPathBuf;

use crate::{
    cli::EnvMode,
    package_json::PackageJson,
    task_graph::{TaskDefinition, TaskOutputs},
};

pub struct PackageTask {
    pub task_id: String,
    task: String,
    package_name: String,
    pkg: PackageJson,
    env_mode: EnvMode,
    pub(crate) task_definition: TaskDefinition,
    pub dir: AbsoluteSystemPathBuf,
    command: String,
    outputs: Vec<String>,
    excluded_outputs: Vec<String>,
    pub(crate) log_file: String,
    hash: String,
}

impl PackageTask {
    pub fn hashable_outputs(&self) -> TaskOutputs {
        let mut inclusion_outputs = vec![format!(".turbo/turbo-{}.log", self.task)];
        inclusion_outputs.extend_from_slice(&self.task_definition.outputs.inclusions[..]);

        let mut hashable = TaskOutputs {
            inclusions: inclusion_outputs,
            exclusions: self.task_definition.outputs.exclusions.clone(),
        };

        hashable.inclusions.sort();
        hashable.exclusions.sort();

        hashable
    }
}
